### PR TITLE
Implement reconciliation of category for create/update/delete categor…

### LINKED ISSUE #4: Base controller and category reconciliation
### What problem are you facing?
After we created category schema is good time to try to do base provider controller, which will use Magento provider config for establishing client for connecting Magento rest api

### How could Crossplane help solve your problem?
- make provider config for Magento connecting credentials 
- make controller which will take provider config from category provider config ref field and will make connection client
- make reconciliation of category for create/update/delete categories.

### PR DESCRIPTION
**Pull request by ca7alindev**

### CODE REVIEW

File: internal/controller/category/category.go [hunk 1]
```diff
@@ -19,7 +19,11 @@ package category
 import (
 	"context"
 	"fmt"
+	"strconv"
 
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/hermsi1337/go-magento2/pkg/magento2/api"
+	"github.com/hermsi1337/go-magento2/pkg/magento2/categories"
```
Comment (evgheni7) [inline internal/controller/category/category.go]: @ca7alindev No,no. This not good, don't use not verified packages. This is no pet project just for fun. Please use pure API calls.

File: apis/category/v1alpha1/category_types.go [hunk 2]
```diff
@@ -50,13 +50,7 @@ type CategoryParameters struct {
 
 // CategoryObservation are the observable fields of a Category.
 type CategoryObservation struct {
-	ID           int    `json:"id,omitempty"`
-	ParentID     int    `json:"parent_id,omitempty"`
-	Name         string `json:"name,omitempty"`
-	IsActive     bool   `json:"is_active,omitempty"`
-	Position     int    `json:"position,omitempty"`
-	Level        int    `json:"level,omitempty"`
-	ProductCount int    `json:"product_count,omitempty"`
```
Comment (evgheni7) [inline apis/category/v1alpha1/category_types.go]: @ca7alindev Why you removed this properties.

**Review by evgheni7 — CHANGES_REQUESTED** [review]
@ca7alindev Rework it using pure REST client.

**Comment by evgheni7** [comment]
@ca7alindev any news?

File: internal/client/categories/categories.go
Comment (evgheni7) [inline internal/client/categories/categories.go:1]: @ca7alindev Please add functions descriptions

File: internal/client/categories/categories.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+	requestBody := map[string]interface{}{
+		"category": observed.Spec.ForProvider,
+	}
+
+	resp, err := restyClient.R().
+		SetHeader("Authorization", "Bearer "+c.AccessToken).
+		SetHeader("Content-Type", "application/json").
+		SetBody(requestBody).
+		Put(path)
```
Comment (evgheni7) [inline internal/client/categories/categories.go]: @ca7alindev Make function for rest client, do not duplicate code.

File: internal/controller/category/category.go [hunk 4]
```diff
@@ -137,22 +149,22 @@ func (c *external) Observe(ctx context.Context, mg resource.Managed) (managed.Ex
 		return managed.ExternalObservation{}, errors.New(errNotCategory)
 	}
 
-	// These fmt statements should be removed in the real implementation.
 	fmt.Printf("Observing: %+v", cr)
```
Comment (evgheni7) [inline internal/controller/category/category.go]: @ca7alindev Remove prints like this, or add debug flag.

File: package/crds/category.magento.web7.md_categories.yaml
Comment (evgheni7) [inline package/crds/category.magento.web7.md_categories.yaml:1]: @ca7alindev Please unify name of CRD files, you mentioned "category" twice.
Shall be magento.web7.md_categories.yaml

File: internal/controller/category/category.go [hunk 5]
```diff
@@ -137,22 +149,22 @@ func (c *external) Observe(ctx context.Context, mg resource.Managed) (managed.Ex
...
+		return managed.ExternalObservation{
+			ResourceExists: false,
+		}, err
+	}
 
+	if desired != nil {
+		cr.Status.SetConditions(xpv1.Available())
+	}
+
+	isUpToDate, _ := categories.IsCategoryUpToDate(cr.Spec.ForProvider.Name, cr, desired)
```
Comment (evgheni7) [inline internal/controller/category/category.go]: @ca7alindev Bad idea to use name for connect resources, in crossplane "external-name" annotation used, check it please how is done in other providers.

File: internal/client/categories/categories.go [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+	var newCategory *v1alpha1.Category
+	err = json.Unmarshal(resp.Body(), &newCategory)
+	if err != nil {
+		return nil, errors.New("failed to parse response: " + err.Error())
+	}
+
+	return newCategory, nil
+}
+
+func UpdateCategoryByName(c *magento.Client, categoryName string, observed *v1alpha1.Category) error {
```
Comment (evgheni7) [inline internal/client/categories/categories.go]: @ca7alindev Not sure about name, shall be used original identification, what is ID for RESTful API

File: examples/sample/category.yaml [hunk 7]
```diff
@@ -0,0 +1,11 @@
+apiVersion: category.magento.web7.md/v1alpha1
```
Comment (evgheni7) [inline examples/sample/category.yaml]: @ca7alindev 
```
apiVersion:  magento.web7.md/v1alpha1
kind: Category

File: examples/sample/category.yaml [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+kind: Category
+metadata:
+  name: test-category
+spec:
+  forProvider:
+    name: "Test Category"
+    is_active: true
+    include_in_menu: true
```
Comment (evgheni7) [inline examples/sample/category.yaml]: @ca7alindev No underscores

File: apis/category/v1alpha1/category_types.go
Comment (evgheni7) [inline apis/category/v1alpha1/category_types.go:1]: @ca7alindev Rename file to simple apis/v1alpha1/category_types.go
Why we need category word twice here?

File: internal/client/client.go [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+// NewClient initializes a new Magento API client configuration
+func NewClient(baseURL, accessToken string) *Client {
+	return &Client{
+		BaseURL:     baseURL,
+		AccessToken: accessToken,
+	}
+}
+
+// CreateRestyClient creates a new resty client with configuration from Client
+func (c *Client) CreateRestyClient() (*resty.Client, error) {
```
Comment (evgheni7) [inline internal/client/client.go]: @ca7alindev Simple Create() don't use customization of methods. This method simple create client.

File: internal/client/categories/categories.go [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+	category, err := GetCategoryByName(c, categoryName)
+	if err != nil {
+		return err
+	}
+	path := fmt.Sprintf("%s/rest/all/V1/categories/%d", c.BaseURL, category.Spec.ForProvider.ID)
```
Comment (evgheni7) [inline internal/client/categories/categories.go]: @ca7alindev Use constant for URLs, you duplicate it many times.

File: internal/client/categories/categories.go [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+type Client struct {
+	magento.Client
+}
+
+func GetCategoryByName(c *magento.Client, categoryName string) (*v1alpha1.Category, error) {
+	path := fmt.Sprintf("%s/rest/all/V1/categories/list", c.BaseURL)
+
+	queryParams := map[string]string{
+		"searchCriteria[filterGroups][0][filters][0][field]": "name",
+		"searchCriteria[filterGroups][0][filters][0][value]": categoryName,
```
Comment (evgheni7) [inline internal/client/categories/categories.go]: @ca7alindev Bad idea, category identifier is ID. Clear is looks like not usable, but you have to keep logic of API, not think like abstract user of your API. Keep standards, that's what is waiting from you by user.

**Review by evgheni7 — CHANGES_REQUESTED** [review]

File: apis/category/v1alpha1/category_types.go
Comment (ca7alindev) [reply apis/category/v1alpha1/category_types.go:1]: @evghen1 in apis directory already exists v1alpha1 where are all magento provider level apis

**Review by evgheni7 — APPROVED** [review]

**Pull request merged**